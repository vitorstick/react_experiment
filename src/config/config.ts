export const CONFIG = {
    URL: {
        SCHEDULES: 'https://fake-schedules-api.herokuapp.com/schedules',
        SCHEDULE_LOGS: 'https://fake-schedules-api.herokuapp.com/scheduleLogs',
    },
};
