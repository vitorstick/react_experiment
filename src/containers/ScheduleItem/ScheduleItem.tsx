import React from 'react';
import {ISchedule} from '../../da/schedule.interface';
import Schedule from '../../components/Schedule/Schedule';

interface Props {
    schedule: ISchedule;
    onScheduleRetire?: (id: number, updatedUser: ISchedule) => void;
    onSelectSchedule?: (schedule: ISchedule) => void;
    onRemoveSchedule?: (id: number, updatedUser: ISchedule) => void;
}

const ScheduleItem: React.FC<Props> = (props) => {
    const retireItem = (retired: boolean): void => {
        const updatedSchedule: ISchedule = {
            ...props.schedule,
            isRetired: retired,
        };
        props.onScheduleRetire(updatedSchedule.id, updatedSchedule);
    };

    const removeItem = (schedule: ISchedule): void => {
        console.log('remove item :: ');
        props.onRemoveSchedule(schedule.id, schedule);
    };

    const selectItem = (schedule: ISchedule): void => {
        props.onSelectSchedule(schedule);
    };

    return (
        <Schedule
            schedule={props.schedule}
            onScheduleRetire={retireItem}
            onSelectSchedule={selectItem}
            onRemoveSchedule={removeItem}
        />
    );
};

export default ScheduleItem;
