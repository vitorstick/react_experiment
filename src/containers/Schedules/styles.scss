.schedules {
    display: flex;
    flex-direction: row;

    &__container {
        display: flex;
        flex-direction: column;
        padding: 0 8px;
        max-height: 85vh;
        overflow-y: scroll;
        scroll-behavior: smooth;
    }

    &__item {
        margin-bottom: 16px;
    }

    .logs__container {
        width: 100%;
    }
}
