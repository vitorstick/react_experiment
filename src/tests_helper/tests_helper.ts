import {ISchedule} from '../da/schedule.interface';

export class TestsMockHelper {
    static getSchedules(): ISchedule[] {
        return [
            {
                id: 19494389,
                name: 'Random Schedule Name (0.8305218495287672)',
                description: 'in dolore aute',
                isRetired: false,
                tasksCount: 9,
                startPoint: '2021-08-12T02:13:39.334Z',
                endPoint: '2021-11-10T02:12:56.326Z',
                dayOfWeek: 1,
                dayOfMonth: 5,
                startDate: '2021-07-03T07:07:18.368Z',
                endDate: '2021-02-05T01:15:54.280Z',
                timePeriod: 12,
                intervalType: 'Minute',
            },
            {
                id: 38792950,
                name: 'Random Schedule Name (0.24529949140144192)',
                description: 'sint ex Lorem Ut incididunt',
                isRetired: false,
                tasksCount: 2,
                startPoint: '2021-12-25T20:34:17.812Z',
                endPoint: '2021-12-13T07:55:34.773Z',
                dayOfWeek: 1,
                dayOfMonth: 24,
                startDate: '2021-09-07T13:27:46.254Z',
                endDate: '2021-02-02T08:05:15.483Z',
                timePeriod: 13,
            },
        ];
    }

    static getSchedule(): ISchedule {
        return this.getSchedules()[0];
    }
}
