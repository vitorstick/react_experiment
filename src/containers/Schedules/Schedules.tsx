import React, {Fragment, ReactElement, useEffect, useState} from 'react';
import ScheduleItem from '../ScheduleItem/ScheduleItem';
import {ISchedule} from '../../da/schedule.interface';
import {CONFIG} from '../../config/config';
import './styles.scss';
import Logs from '../Logs/Logs';

const Schedules: React.FC = () => {
    const [items, setItems] = useState([]);
    const [logs, setLogs] = useState([]);
    const [error, setError] = useState(null);
    const [loading, setIsLoading] = useState(true);
    const [editing, setEditing] = useState(false);
    const [removing, setRemoving] = useState(false);

    const updateItem = (id: number, updatedItem: ISchedule) => {
        setEditing(true);

        setItems(items.map((item) => (item.id === id ? updatedItem : item)));
        setEditing(false);
    };

    const removeItem = (id: number, removedItem: ISchedule) => {
        setRemoving(true);

        setItems(items.filter((item) => item.id !== id));
        setRemoving(false);
    };

    const selectItem = (schedule: ISchedule) => {
        fetch(`${CONFIG.URL.SCHEDULE_LOGS}?scheduleId=${schedule.id}`)
            .then((res) => {
                return res.json();
            })
            .then(
                (result: ISchedule[]) => {
                    console.log('logs :: ', result);
                    setLogs(result);
                },
                (error) => {
                    setError(error);
                },
            )
            .finally(() => setIsLoading(false));
    };

    useEffect(() => {
        fetch(`${CONFIG.URL.SCHEDULES}`)
            .then((res) => {
                return res.json();
            })
            .then(
                (result: ISchedule[]) => {
                    setItems(result);
                },
                (error) => {
                    console.log('error :: ', error);
                    setError(error);
                },
            )
            .finally(() => setIsLoading(false));
    }, []);

    if (error) {
        return <h2>Error</h2>;
    }

    return (
        <div className="schedules">
            <div className="schedules__container">
                {items?.map((item: ISchedule): ReactElement => {
                    return (
                        <div className="schedules__item" key={item.id}>
                            {editing || removing || loading ? (
                                <Fragment>
                                    <div className="schedules__loading">
                                        Loading
                                    </div>
                                </Fragment>
                            ) : (
                                <Fragment>
                                    <ScheduleItem
                                        schedule={item}
                                        onScheduleRetire={updateItem}
                                        onSelectSchedule={selectItem}
                                        onRemoveSchedule={removeItem}
                                    />
                                </Fragment>
                            )}
                        </div>
                    );
                })}
            </div>
            <div className="logs__container">
                <Logs logs={logs} />
            </div>
        </div>
    );
};

export default Schedules;
