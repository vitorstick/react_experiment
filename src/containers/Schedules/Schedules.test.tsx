import React from 'react';
import 'jsdom-global/register';
import fetchMock from 'jest-fetch-mock';
import {mount, ReactWrapper, shallow} from 'enzyme';
import Schedules from './Schedules';
import ScheduleItem from '../ScheduleItem/ScheduleItem';
import fetch from 'node-fetch';
import {CONFIG} from '../../config/config';
import {act} from 'react-dom/test-utils';
import {render, unmountComponentAtNode} from 'react-dom';
import {TestsMockHelper} from '../../tests_helper/tests_helper';
import {ISchedule} from '../../da/schedule.interface';

describe('<Schedules />', () => {
    let component;
    const schedules: ISchedule[] = TestsMockHelper.getSchedules();

    beforeEach(() => {
        component = shallow(<Schedules />);
        fetchMock.doMock();
    });

    test('Should mount component', () => {
        expect(component).toMatchSnapshot();
        expect(component.length).toBe(1);
    });

    test('should call API and returns data to me', async () => {
        const container = document.createElement('div');
        document.body.appendChild(container);

        const testData = schedules;
        const response = new Response();
        jest.spyOn(global, 'fetch').mockImplementation(() =>
            Promise.resolve(response),
        );

        await act(async () => {
            render(<Schedules />, container);
        });

        expect(container).toBeDefined();

        global.fetch.mockRestore();
    });
});
