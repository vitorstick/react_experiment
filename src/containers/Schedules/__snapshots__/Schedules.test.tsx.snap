// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Schedules /> Should mount component 1`] = `
<h2>
  Loading
</h2>
`;

exports[`<Schedules /> Snapshot 1`] = `
<h2>
  Loading
</h2>
`;
